packages/node/src/node/client/ClientNodes.ts

```diff
@@ -107,6 +119,90 @@ export class ClientNodes extends EndpointContainer<ClientNode> {
 
         super.add(node);
     }
+
+    override async close() {
+        this.#cancelExpiration();
+        await super.close();
+    }
+
+    #cancelExpiration() {
+        if (this.#expirationInterval) {
+            this.#expirationInterval.cancel();
+            this.#expirationInterval = undefined;
+        }
+    }
+
+    #manageExpiration() {
+        if (this.#expirationInterval) {
+            if (!this.size) {
+                this.#cancelExpiration();
+            }
+            return;
+        }
+
+        if (!this.size) {
+            return;
+        }
+
+        this.#expirationInterval = Time.wait("client node expiration", EXPIRATION_INTERVAL).then(async () => {
+            this.#expirationInterval = undefined;
+            try {
+                await this.#cullExpiredNodesAndAddresses();
+            } finally {
+                this.#manageExpiration();
+            }
+        });
+    }
+
+    async #cullExpiredNodesAndAddresses() {
+        const now = Time.nowMs();
+
+        for (const node of this) {
+            const state = node.state.commissioning;
+            const { addresses } = state;
+            const isCommissioned = state.peerAddress !== undefined;
+
+            // Shortcut for conditions we know no change is possible
+            if (addresses === undefined || (isCommissioned && addresses.length === 1)) {
+                return;
+            }
+
+            // Remove expired addresses
+            let newAddresses = addresses.filter(addr => {
+                const exp = expirationOf(addr);
+                if (exp === undefined) {
+                    return true;
+                }
+
+                return exp > now;
+            });
+
+            // Cull commissionable nodes that have expired
+            if (!isCommissioned) {
+                if (!newAddresses?.length || expirationOf(state) <= now) {
+                    await node.delete();
+                    continue;
+                }
+            }
+
+            // If the node is commissioned, do not remove the last address.  Instead keep the "least expired" addresses
+            if (isCommissioned && addresses.length && !newAddresses.length) {
```

Comment:
Ideally use the "currently connected address" because in fact it could be that others get reported, but the currently connected one is still potentially better then any other because validated to work. So the currently connected should take precedence

Reply:
For commissioned nodes we actually currently only have the one address because PeerSet doesn't publish the other addresses.  When we resolve that we should put the current address at the top of the list so it has precedence, but not relevant right now